File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
         stack<char> c;

    for(int i=0; i<s.length(); i++){

        if(s[i] == '(' || s[i] == '[' || s[i] == '{'){
            c.push(s[i]);
        }

        else if(s[i] == ')' || s[i] == ']' || s[i] == '}'){

            if(c.empty()){
                return false;
            }

            char top = c.top();
            c.pop();

            if((s[i] == ')' && top != '(') ||
               (s[i] == '}' && top != '{') ||
               (s[i] == ']' && top != '[')) {

                return false;
            }
        }
    }

    return c.empty();
        
    }
};

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
          int n = s.length();
        Stack<Character> st = new Stack<>();
        for(int i =0; i<n; i++){
            if(!st.empty() && 
            (((st.peek() == '(') && (s.charAt(i) == ')')) || 
            ((st.peek() == '[') && (s.charAt(i) == ']')) ||
            ((st.peek() == '{') && (s.charAt(i) == '}'))))
             {
                st.pop();
            }
            else{
                st.push(s.charAt(i));
            }
        }       
        return st.empty();
    }
}